# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, balanced_accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [3]:
data

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0                0.00            0.00  ...        0.000        0.000   
1                0.00            0.94  ...        0.000        0.132   
2                0.64            0.25  ...        0.010        0.143   
3                0.31            0.63  ...        0.000        0.137   
4                0.31            0.63  ...        0.000        0.135   
...               ...             ...  ...          ...          ...   
4596             0.00            0.00  ...        0.000        0.232   
4597             0.00            0.00  ...        0.000        0.000   
4598             0.00            0.00  ...        0.102        0.718   
4599             0.00            0.00  ...        0.000        0.057   
4600             0.00            0.00  ...        0.000        0.000   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0             0.0        0.778        0.000        0.000   
1             0.0        0.372        0.180        0.048   
2             0.0        0.276        0.184        0.010   
3             0.0        0.137        0.000        0.000   
4             0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...   
4596          0.0        0.000        0.000        0.000   
4597          0.0        0.353        0.000        0.000   
4598          0.0        0.000        0.000        0.000   
4599          0.0        0.000        0.000        0.000   
4600          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537                          40   
...                          ...                         ...   
4596                       1.142                           3   
4597                       1.555                           4   
4598                       1.404                      

In [4]:
# Check column defnitions for categorical values.
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [5]:
#Check Summary Statistics
data.describe()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.104553           0.213015       0.280656      0.065425   
std          0.305358           1.290575       0.504143      1.395151   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.000000       0.420000      0.000000   
max          4.540000          14.280000       5.100000     42.810000   

       word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
count    4601.000000     4601.000000       4601.000000         4601.000000   
mean        0.312223        0.095901          0.114208            0.105295   
std         0.672513        0.273824          0.391441            0.401071   
min         0.000000        0.000000          0.000000            0.000000   
25%         0.000000        0.000000          0.000000            0.000000   
50%         0.000000        0.000000          0.000000            0.000000   
75%         0.380000        0.000000          0.000000            0.000000   
max        10.000000        5.880000          7.270000           11.110000   

       word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
count      4601.000000     4601.000000  ...  4601.000000  4601.000000   
mean          0.090067        0.239413  ...     0.038575     0.139030   
std           0.278616        0.644755  ...     0.243471     0.270355   
min           0.000000        0.000000  ...     0.000000     0.000000   
25%           0.000000        0.000000  ...     0.000000     0.000000   
50%           0.000000        0.000000  ...     0.000000     0.065000   
75%           0.000000        0.160000  ...     0.000000     0.188000   
max           5.260000       18.180000  ...     4.385000     9.752000   

       char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
count  4601.000000  4601.000000  4601.000000  4601.000000   
mean      0.016976     0.269071     0.075811     0.044238   
std       0.109394     0.815672     0.245882     0.429342   
min       0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.315000     0.052000     0.000000   
max       4.081000    32.478000     6.003000    19.829000   

       capital_run_length_average  capital_run_length_longest  \
count                 4601.000000                 4601.000000   
mean                     5.191515                   52.172789   
std                     31.729449                  194.891310   
min                      1.000000                    1.000000   
25%                      1.588000                    6.000000   
50%                      2.276000                   15.000000   
75%                      3.706000                   43.000000   
max                   1102.500000                 9989.000000   

       capital_run_length_total         spam  
count               4601.000000  4601.000000  
mean                 283.289285     0.394045  
std                  606.347851     0.488698  
min                    1.000000     0.000000  
25%                   35.000000     0.000000  
50%                   95.000000     0.000000  
75%                  266.000000     1.000000  
max                15841.000000     1.000000  

[8 rows x 58 columns]

In [6]:
#Check Value distribution of the target column
data.spam.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Random Forests Classifier usually performs better. But because this dataset appears well defined, without any "unwanted" noise or big outliers, Logistic Regression might outperform Random Forest for this dataset. But it needs to be proven by running both the models under similar conditions.

## Split the Data into Training and Testing Sets

In [7]:
# Create the labels set `y` and features DataFrame `X`
# Features
X=data.copy().drop(columns='spam')

# Target
y=data['spam']

In [8]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

given both value counts are close, it can be treated as a Balanced dataset.

In [9]:
numerical = X.select_dtypes(include=['int64','float64']).columns.tolist()
numerical

['word_freq_make',
 'word_freq_address',
 'word_freq_all',
 'word_freq_3d',
 'word_freq_our',
 'word_freq_over',
 'word_freq_remove',
 'word_freq_internet',
 'word_freq_order',
 'word_freq_mail',
 'word_freq_receive',
 'word_freq_will',
 'word_freq_people',
 'word_freq_report',
 'word_freq_addresses',
 'word_freq_free',
 'word_freq_business',
 'word_freq_email',
 'word_freq_you',
 'word_freq_credit',
 'word_freq_your',
 'word_freq_font',
 'word_freq_000',
 'word_freq_money',
 'word_freq_hp',
 'word_freq_hpl',
 'word_freq_george',
 'word_freq_650',
 'word_freq_lab',
 'word_freq_labs',
 'word_freq_telnet',
 'word_freq_857',
 'word_freq_data',
 'word_freq_415',
 'word_freq_85',
 'word_freq_technology',
 'word_freq_1999',
 'word_freq_parts',
 'word_freq_pm',
 'word_freq_direct',
 'word_freq_cs',
 'word_freq_meeting',
 'word_freq_original',
 'word_freq_project',
 'word_freq_re',
 'word_freq_edu',
 'word_freq_table',
 'word_freq_conference',
 'char_freq_;',
 'char_freq_(',
 'char_freq_[',
 '

In [10]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4576            0.00                0.0           0.00           0.0   
4401            0.00                0.0           0.00           0.0   
3707            0.17                0.0           0.17           0.0   
2362            0.00                0.0           0.00           0.0   
1537            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4576           0.00             0.0               0.0                 0.0   
4401           0.00             0.0               0.0                 0.0   
3707           0.00             0.0               0.0                 0.0   
2362           0.00             0.0               0.0                 0.0   
1537           2.17             0.0               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4576              0.0             0.0  ...                   0.0        0.000   
4401              0.0             0.0  ...                   0.0        0.000   
3707              0.8             0.0  ...                   0.0        0.253   
2362              0.0             0.0  ...                   0.0        0.000   
1537              0.0             0.0  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4576        0.131        0.000          0.0        0.000          0.0   
4401        0.000        0.000          0.0        0.000          0.0   
3707        0.168        0.084          0.0        0.024          0.0   
2362        0.000        0.000          0.0        0.000          0.0   
1537        0.000        0.000          0.0        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
4576                       1.488                           5   
4401                       1.571                           5   
3707                       4.665                          81   
2362                       4.228                          53   
1537                       1.333                           5   

      capital_run_length_total  
4576                        64  
4401                        11  
3707                      1031  
2362                       148  
1537                        16  

[5 rows x 57 columns]

In [11]:
X_test.tail()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1089            0.34               0.42           0.25           0.0   
929             0.00               0.81           1.62           0.0   
1545            0.41               0.00           0.41           0.0   
4356            0.00               0.00           0.00           0.0   
936             0.00               0.47           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1089           0.08            0.42              0.08                0.25   
929            2.43            0.00              0.00                0.00   
1545           0.00            0.00              0.41                0.00   
4356           0.00            0.00              0.00                0.00   
936            0.94            0.00              0.94                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1089             0.08            1.62  ...                   0.0          0.0   
929              0.00            0.81  ...                   0.0          0.0   
1545             0.00            0.00  ...                   0.0          0.0   
4356             0.00            0.00  ...                   0.0          0.0   
936              0.00            0.00  ...                   0.0          0.0   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1089        0.065        0.000        0.261        0.294        0.065   
929         0.121        0.000        0.121        0.000        0.000   
1545        0.067        0.000        0.067        0.000        0.000   
4356        0.775        0.000        0.000        0.000        0.000   
936         0.074        0.074        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1089                       3.282                          62   
929                        4.035                          43   
1545                       1.863                          14   
4356                       1.000                           1   
936                        2.125                          11   

      capital_run_length_total  
1089                       535  
929                        113  
1545                        41  
4356                         5  
936                        102  

[5 rows x 57 columns]

In [12]:
print(f"Train Shape: {X_train.shape,y_train.shape}")
print(f"Test Shape: {X_test.shape,y_test.shape}")

Train Shape: ((3450, 57), (3450,))
Test Shape: ((1151, 57), (1151,))


## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [13]:
from sklearn.preprocessing import StandardScaler
# Create the StandardScaler instance
X_scaler = StandardScaler()

In [14]:
# Fit the Standard Scaler with the training data
X_scaler.fit(X_train)

StandardScaler()

In [15]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [16]:
X_train_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ..., -0.13974678,
        -0.23023328, -0.37548424],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.13640572,
        -0.23023328, -0.46466704],
       [ 0.21317665, -0.16744248, -0.22391235, ..., -0.01186062,
         0.13008597,  1.25168127],
       ...,
       [-0.35811925, -0.16744248,  0.94500764, ..., -0.10138496,
        -0.16385868, -0.06418577],
       [-0.35811925, -0.16744248,  1.16294188, ..., -0.1351981 ,
        -0.20178702, -0.01707033],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15524447,
        -0.23971537, -0.42933046]])

In [17]:
X_test_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ...,  0.07078032,
        -0.09748408, -0.12139738],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.11736569,
        -0.14963555, -0.32836879],
       [-0.35811925,  0.24309223, -0.56071981, ..., -0.01628853,
        -0.01214531, -0.27620527],
       ...,
       [ 1.01971205, -0.16744248,  0.25158053, ..., -0.12465162,
        -0.1875639 , -0.41418621],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15939061,
        -0.24919745, -0.47476321],
       [-0.35811925,  0.1966166 , -0.56071981, ..., -0.11410514,
        -0.20178702, -0.31154185]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [18]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state = 1)

In [19]:
# Train the model using training dataset
lr.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [20]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_predictions = lr.predict(X_test_scaled)
# Review the predictions
lr_predictions

array([0, 0, 1, ..., 0, 0, 1], dtype=int64)

In [21]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"Logistics Regression Accuracy Score: {accuracy_score(y_test, lr_predictions)}")

Logistics Regression Accuracy Score: 0.9287576020851434


In [22]:
# Calculate the Balanced accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"Logistics Regression Balanced Accuracy Score: {balanced_accuracy_score(y_test, lr_predictions)}")

Logistics Regression Balanced Accuracy Score: 0.9188349976224441


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [23]:
##### Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=128, random_state=1)
rf.fit(X_train_scaled, y_train)

RandomForestClassifier(n_estimators=128, random_state=1)

In [24]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_predictions = rf.predict(X_test_scaled)

# Review the predictions
rf_predictions

array([1, 1, 1, ..., 1, 0, 1], dtype=int64)

In [25]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.


In [26]:
print(f"Random Forest Accuracy Score: {accuracy_score(y_test, rf_predictions)}")

Random Forest Accuracy Score: 0.9669852302345786


In [27]:
# Calculate the Balanced accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"Random Forest Balanced Accuracy Score: {balanced_accuracy_score(y_test, rf_predictions)}")

Random Forest Balanced Accuracy Score: 0.9641432873672531


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

Random Forest classifier has predicted results with slightly higher accuracy: 96% in Random Forest to 93% in Logistics Regression.

Results are not surprising. I expected Logistics regression to have much closer accuracy (within 1%) given the type of data.